src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseDataFileService.cs

```diff
@@ -184,26 +185,69 @@ await ids.Select(id => _releaseFileRepository.CheckFileExists(releaseId, id, Fil
 
                     await filesExcludingReplacements
                         .ToAsyncEnumerable()
+                        .OrderBy(releaseFile => releaseFile.Order)
+                        .ThenBy(releaseFile => releaseFile.Name)
                         .ForEachAwaitAsync(async file =>
                             fileList.Add(await GetDataFileInfo(releaseId, file.File)));
 
                     return fileList
-                        .OrderBy(file => file.Name)
                         .AsEnumerable();
                 });
         }
 
+        public async Task<Either<ActionResult, IEnumerable<DataFileInfo>>> ReorderDataFiles(
+            Guid releaseId,
+            Dictionary<Guid, int> newOrder)
+        {
+            return await _persistenceHelper
+                .CheckEntityExists<Release>(releaseId)
+                .OnSuccess(release => _userService.CheckCanUpdateRelease(release))
+                .OnSuccess(async _ =>
+                {
+                    var releaseFiles = await _contentDbContext.ReleaseFiles
+                        .Include(releaseFile => releaseFile.File)
+                        .Where(releaseFile =>
+                            releaseFile.File.Type == FileType.Data
+                            && releaseFile.ReleaseId == releaseId
+                            && !releaseFile.File.ReplacingId.HasValue)
+                        .ToListAsync();
```

Comment:
You could use `ToDictionary` here to provide a quicker lookup of the files by id because the way `releaseFiles.Find` is used below is iterating over all the files to find the correct one by id, and that's happening for every file within the loop.